File: course_schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        requirements = {i: [] for i in range(numCourses)}

        for course, neighbor in prerequisites:
            requirements[course].append(neighbor)

        path = {}

        def dfs(course):
            if course in path and path[course] == True:
                return False
            if requirements[course] == []:
                return True

            path[course] = True
            for neighbor in requirements[course]:
                if not dfs(neighbor):
                    return False
            path[course] = False
            requirements[course] = []

            return True

        for c in range(numCourses):
            if not dfs(c):
                return False

        return True

File: courseSchedule.js
class Solution {
  /**
   * @param {number} numCourses
   * @param {number[][]} prerequisites
   * @return {boolean}
   */
  canFinish(numCourses, prerequisites) {
    let requirements = {};
    for (let i = 0; i < numCourses; i++) {
      requirements[i] = [];
    }

    for (let [course, prerequisite] of prerequisites) {
      requirements[course].push(prerequisite);
    }

    let path = {};

    const dfs = (course) => {
      if (course in path && path[course] === true) {
        return false;
      }

      if (requirements[course].length === 0) {
        return true;
      }

      path[course] = true;
      for (let prerequisite of requirements[course]) {
        if (!dfs(prerequisite)) {
          return false;
        }
      }
      path[course] = false;
      requirements[course] = [];

      return true;
    };

    for (let c = 0; c < numCourses; c++) {
      if (!dfs(c)) {
        return false;
      }
    }

    return true;
  }
}
